Slot typings

**Is your feature request related to a problem? Please describe.**

Volar is capable of checking the slot props type and can assert if the template of a component use them properly.
Until now, you had to provide these types manually, with something like this for vee-validate:

```ts
import { Field, Form, useField, useForm } from 'vee-validate';
import type { UnwrapRef } from 'vue';

declare module '@vue/runtime-core' {
  interface GlobalComponents {
    Field: typeof Field & {
      __VLS_slots: {
        default: UnwrapRef<ReturnType<typeof useField>> & { field: any };
      };
    };
    Form: typeof Form & {
      __VLS_slots: {
        default: UnwrapRef<ReturnType<typeof useForm>>;
      };
    };
  }
}
```

This works great, allowing to catch mistakes in a template:

```vue
<!-- metas instead of meta -->
<Field name="login" rules="required|min:3" v-slot="{ field, metas }">
```

is caught by Volar:
```bash
$ vue-tsc --noEmit
src/views/Register.vue:11:69 - error TS2339: Property 'metas' does not exist on type '{ label?: string | undefined; meta: { touched: boolean; dirty: boolean; valid: boolean; validated: boolean; pending: boolean; initialValue?: unknown; }; name: string; type?: string | undefined; value: unknown; ... 14 more ...; setValue: (value: unknown) => void; } & { ...; }'.

11         <Field name="login" rules="required|min:3" v-slot="{ field, metas }">
                                                                       ~~~~~

Found 1 error.
```

**Describe the solution you'd like**

Volar now can pick pick up `$slots` definition if provided since https://github.com/johnsoncodehk/volar/blob/master/CHANGELOG.md#02728

`vue-router` added them in its latest release (see https://github.com/vuejs/vue-router-next/commit/1b28dd428c691e4f454c9731d5bdbebc737e257f), allowing to simplify the declaration of the Router components from:

```ts
  interface GlobalComponents {
    RouterLink: typeof RouterLink & {
      __VLS_slots: {
        default: UnwrapRef<ReturnType<typeof useLink>>;
      };
    };
    RouterView: typeof RouterView & {
      __VLS_slots: {
        default: {
          Component: VNode;
          route: RouteLocationNormalized & { href: string };
        };
      };
    };
```

to 
```ts
  interface GlobalComponents {
    RouterLink: typeof RouterLink;
    RouterView: typeof RouterView;
```

It would be awesome to provide slots types out of the box for vee-validate Form and Field, allowing developers to remove the custom declaration from their applications.

**Describe alternatives you've considered**
It's possible to manually declare the slot types, but it is cumbersome to do it in every application, and error-prone.

--- a/packages/vee-validate/src/ErrorMessage.ts
+++ b/packages/vee-validate/src/ErrorMessage.ts
@@ -1,8 +1,12 @@
-import { inject, h, defineComponent, computed, resolveDynamicComponent } from 'vue';
+import { inject, h, defineComponent, computed, resolveDynamicComponent, VNode } from 'vue';
 import { FormContextKey } from './symbols';
 import { normalizeChildren } from './utils';
 
-export const ErrorMessage = defineComponent({
+interface ErrorMessageSlotProps {
+  message: string | undefined;
+}
+
+const ErrorMessageImpl = defineComponent({
   name: 'ErrorMessage',
   props: {
     as: {
@@ -20,7 +24,7 @@ export const ErrorMessage = defineComponent({
       return form?.errors.value[props.name];
     });
 
-    function slotProps() {
+    function slotProps(): ErrorMessageSlotProps {
       return {
         message: message.value,
       };
@@ -33,7 +37,7 @@ export const ErrorMessage = defineComponent({
       }
 
       const tag = (props.as ? resolveDynamicComponent(props.as) : props.as) as string;
-      const children = normalizeChildren(tag, ctx, slotProps);
+      const children = normalizeChildren(tag, ctx, slotProps as any);
 
       const attrs = {
         role: 'alert',
@@ -56,3 +60,11 @@ export const ErrorMessage = defineComponent({
     };
   },
 });
+
+export const ErrorMessage = ErrorMessageImpl as typeof ErrorMessageImpl & {
+  new (): {
+    $slots: {
+      default: (arg: ErrorMessageSlotProps) => VNode[];
+    };
+  };
+};

--- a/packages/vee-validate/src/Field.ts
+++ b/packages/vee-validate/src/Field.ts
@@ -1,9 +1,21 @@
-import { h, defineComponent, toRef, SetupContext, resolveDynamicComponent, computed, watch, PropType } from 'vue';
+import {
+  h,
+  defineComponent,
+  toRef,
+  SetupContext,
+  resolveDynamicComponent,
+  computed,
+  watch,
+  PropType,
+  VNode,
+} from 'vue';
 import { getConfig } from './config';
 import { RuleExpression, useField } from './useField';
 import { normalizeChildren, hasCheckedAttr, shouldHaveValueBinding, isPropPresent, normalizeEventValue } from './utils';
 import { toNumber } from '../../shared';
 import { IS_ABSENT } from './symbols';
+import { FieldMeta } from './types';
+import { FieldContext } from '.';
 
 interface ValidationTriggersProps {
   validateOnMount: boolean;
@@ -13,7 +25,30 @@ interface ValidationTriggersProps {
   validateOnModelUpdate: boolean;
 }
 
-export const Field = defineComponent({
+interface FieldBindingObject<TValue = unknown> {
+  name: string;
+  onBlur: (e: Event) => unknown;
+  onInput: (e: Event) => unknown;
+  onChange: (e: Event) => unknown;
+  'onUpdate:modelValue'?: ((e: TValue) => unknown) | undefined;
+  value?: unknown;
+  checked?: boolean;
+}
+
+interface FieldSlotProps<TValue = unknown>
+  extends Pick<
+    FieldContext,
+    'validate' | 'resetField' | 'handleChange' | 'handleReset' | 'handleBlur' | 'setTouched' | 'setErrors'
+  > {
+  field: FieldBindingObject<TValue>;
+  value: TValue;
+  meta: FieldMeta<TValue>;
+  errors: string[];
+  errorMessage: string | undefined;
+  handleInput: FieldContext['handleChange'];
+}
+
+const FieldImpl = defineComponent({
   name: 'Field',
   inheritAttrs: false,
   props: {
@@ -135,11 +170,15 @@ export const Field = defineComponent({
     const fieldProps = computed(() => {
       const { validateOnInput, validateOnChange, validateOnBlur, validateOnModelUpdate } =
         resolveValidationTriggers(props);
-      const baseOnBlur = [handleBlur, ctx.attrs.onBlur, validateOnBlur ? validateField : undefined].filter(Boolean);
-      const baseOnInput = [(e: unknown) => onChangeHandler(e, validateOnInput), ctx.attrs.onInput].filter(Boolean);
-      const baseOnChange = [(e: unknown) => onChangeHandler(e, validateOnChange), ctx.attrs.onChange].filter(Boolean);
+      const baseOnBlur: any = [handleBlur, ctx.attrs.onBlur, validateOnBlur ? validateField : undefined].filter(
+        Boolean
+      );
+      const baseOnInput: any = [(e: unknown) => onChangeHandler(e, validateOnInput), ctx.attrs.onInput].filter(Boolean);
+      const baseOnChange: any = [(e: unknown) => onChangeHandler(e, validateOnChange), ctx.attrs.onChange].filter(
+        Boolean
+      );
 
-      const attrs: Record<string, any> = {
+      const attrs: FieldBindingObject<unknown> = {
         name: props.name,
         onBlur: baseOnBlur,
         onInput: baseOnInput,
@@ -147,7 +186,7 @@ export const Field = defineComponent({
       };
 
       if (validateOnModelUpdate) {
-        attrs['onUpdate:modelValue'] = [onChangeHandler];
+        attrs['onUpdate:modelValue'] = [onChangeHandler] as any;
       }
 
       if (hasCheckedAttr(ctx.attrs.type) && checked) {
@@ -177,7 +216,7 @@ export const Field = defineComponent({
       }
     });
 
-    function slotProps() {
+    function slotProps(): FieldSlotProps {
       return {
         field: fieldProps.value,
         value: value.value,
@@ -205,7 +244,7 @@ export const Field = defineComponent({
 
     return () => {
       const tag = resolveDynamicComponent(resolveTag(props, ctx)) as string;
-      const children = normalizeChildren(tag, ctx, slotProps);
+      const children = normalizeChildren(tag, ctx, slotProps as any);
 
       if (tag) {
         return h(
@@ -261,3 +300,11 @@ function resolveInitialValue(props: Record<string, unknown>, ctx: SetupContext<a
 
   return isPropPresent(props, 'modelValue') ? props.modelValue : undefined;
 }
+
+export const Field = FieldImpl as typeof FieldImpl & {
+  new (): {
+    $slots: {
+      default: (arg: FieldSlotProps<unknown>) => VNode[];
+    };
+  };
+};

--- a/packages/vee-validate/src/FieldArray.ts
+++ b/packages/vee-validate/src/FieldArray.ts
@@ -1,8 +1,9 @@
-import { defineComponent, toRef } from 'vue';
+import { defineComponent, toRef, UnwrapRef, VNode } from 'vue';
+import { FieldArrayContext } from './types';
 import { useFieldArray } from './useFieldArray';
 import { normalizeChildren } from './utils';
 
-export const FieldArray = defineComponent({
+const FieldArrayImpl = defineComponent({
   name: 'FieldArray',
   inheritAttrs: false,
   props: {
@@ -44,3 +45,11 @@ export const FieldArray = defineComponent({
     };
   },
 });
+
+export const FieldArray = FieldArrayImpl as typeof FieldArrayImpl & {
+  new (): {
+    $slots: {
+      default: (arg: UnwrapRef<FieldArrayContext>) => VNode[];
+    };
+  };
+};

--- a/packages/vee-validate/src/Form.ts
+++ b/packages/vee-validate/src/Form.ts
@@ -1,9 +1,34 @@
-import { h, defineComponent, toRef, resolveDynamicComponent, PropType } from 'vue';
+import { h, defineComponent, toRef, resolveDynamicComponent, PropType, VNode, UnwrapRef } from 'vue';
 import { useForm } from './useForm';
 import { SubmissionHandler, InvalidSubmissionHandler } from './types';
 import { isEvent, normalizeChildren } from './utils';
+import { FormContext } from '.';
 
-export const Form = defineComponent({
+type FormSlotProps = UnwrapRef<
+  Pick<
+    FormContext,
+    | 'meta'
+    | 'errors'
+    | 'values'
+    | 'isSubmitting'
+    | 'submitCount'
+    | 'validate'
+    | 'validateField'
+    | 'handleReset'
+    | 'submitForm'
+    | 'setErrors'
+    | 'setFieldError'
+    | 'setFieldValue'
+    | 'setValues'
+    | 'setFieldTouched'
+    | 'setTouched'
+    | 'resetForm'
+  >
+> & {
+  handleSubmit: (evt: Event | SubmissionHandler, onSubmit?: SubmissionHandler) => Promise<unknown>;
+};
+
+const FormImpl = defineComponent({
   name: 'Form',
   inheritAttrs: false,
   props: {
@@ -89,7 +114,7 @@ export const Form = defineComponent({
       return handleSubmit(onSuccess as SubmissionHandler<Record<string, unknown>>, props.onInvalidSubmit)(evt as Event);
     }
 
-    function slotProps() {
+    function slotProps(): FormSlotProps {
       return {
         meta: meta.value,
         errors: errors.value,
@@ -127,7 +152,7 @@ export const Form = defineComponent({
     return function renderForm() {
       // avoid resolving the form component as itself
       const tag = props.as === 'form' ? props.as : (resolveDynamicComponent(props.as) as string);
-      const children = normalizeChildren(tag, ctx, slotProps);
+      const children = normalizeChildren(tag, ctx, slotProps as any);
 
       if (!props.as) {
         return children;
@@ -155,3 +180,11 @@ export const Form = defineComponent({
     };
   },
 });
+
+export const Form = FormImpl as typeof FormImpl & {
+  new (): {
+    $slots: {
+      default: (arg: FormSlotProps) => VNode[];
+    };
+  };
+};

